3985 automatic versioning

AndyMender opened the pull request:
Discussed in #3985 

- uses `bumpversion` to update the `__version__` string in `qutebrowser/__init__.py`
- adds an update_version.py dev script to scripts/dev to handle files which require version bumping, but cannot be easily accessed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4304)
<!-- Reviewable:end -->

The-Compiler commented on .bumpversion.cfg:
```diff
@@ -0,0 +1,9 @@
+current_version = 1.5.0
+commit = True
+tag = True
+tag_name = v{new_version}
```
I'd like to have signed tags, but it looks like bumpversion doesn't have that - it's also not maintained anymore from what it seems... According to https://github.com/peritus/bumpversion/issues/169 [bump2version](https://github.com/c4urself/bump2version) is a commonly accepted and maintained fork.

Could you please look into using that instead, so signed tags can be turned on? :slightly_smiling_face:

AndyMender replied:
I'll look into it tomorrow. Thanks for the suggestion :).

AndyMender replied:
`bump2version` is being used in update_version.py now and .bumpversion.cfg was updated with the new `sign_tags` toggle.

The-Compiler commented on qutebrowser/__init__.py:
```diff
@@ -26,8 +26,8 @@
-__version_info__ = (1, 5, 0)
-__version__ = '.'.join(str(e) for e in __version_info__)
+__version__ = "1.5.0"
+__version_info__ = __version__.split('.')
```
Those should be integers, not strings.

AndyMender replied:
Yes! Apologies for the omission!

AndyMender replied:
Fixed with a list comprehension

The-Compiler commented on requirements.txt:
```diff
@@ -1,9 +1,11 @@
 # This file is automatically generated by scripts/dev/recompile_requirements.py
 
 attrs==18.2.0
+bumpversion==0.5.3
```
That'd mean everyone installing qutebrowser also needs to install bumpversion/lxml. Right now there's nothing like `requirements-dev.txt` (only `misc/requirements/*`, but it doesn't really fit anywhere), so I'd propose just removing those.

AndyMender replied:
Removed as suggested :).

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+from datetime import date
```
Please just use `import datetime` (I find it quite difficult to see where stuff is coming from when importing things from modules, so that's usually not done in the qutebrowser codebase :wink:)

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+
+from lxml import etree
+
+# TODO: move to global constants?
```
What do you mean with that? Those do seem to be global constants?

AndyMender replied:
qutebrowser.appdata.xml is a central file and its path could easily be moved to a centrally located store like `qutebrowser/__init__.py` for instance.

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+def bump_version(version_leap = "patch"):
+    """Update qutebrowser release version in .bumpversion.cfg
+
+    :param version_leap: define the jump between versions ("major", "minor", "patch")
```
Please use the google syntax for docstrings (see [this part](https://github.com/qutebrowser/qutebrowser/blob/master/doc/contributing.asciidoc#style-conventions) of the contribution docs; same for other docstrings as well).

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+
+
+def bump_version(version_leap = "patch"):
+    """Update qutebrowser release version in .bumpversion.cfg
```
Please add a period at the end, as per [pep257](https://www.python.org/dev/peps/pep-0257/) (same for toehr docstrings as well).

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+
+    # NOTE: this may rely on the python version used to launch the function
+    # and whether the wrapper script 'bumpversion' was provided
+    subprocess.run(['bumpversion', version_leap, '--allow-dirty'])
```
Please use `check=True` so the script aborts when this failed.

The-Compiler replied:
Probably `sys.executable, '-m', 'bumpversion'` or similar would work better, as it doesn't rely on it being in the `$PATH`.

The-Compiler replied:
Not sure if it makes sense to specify `--allow-dirty` by default - or is that needed because of stuff like the `.tox` folder?

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+    """
+
+    # make sure the correct version jump was chosen
+    if version_leap not in ("major", "minor", "patch"):
```
nitpick: I'd prefer a list to a tuple here.

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+version_xpath = '//*[@type="desktop"]/releases'
+
+
+def bump_version(version_leap = "patch"):
```
No spaces around `=` for default arguments, as per [pep8](https://www.python.org/dev/peps/pep-0008/#other-recommendations).

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+
+    :param version_leap: define the jump between versions ("major", "minor", "patch")
+    """
+
```
No blank lines after docstrings (same for other docstrings as well)

The-Compiler commented on .bumpversion.cfg:
```diff
@@ -0,0 +1,9 @@
+
+[bumpversion:file:qutebrowser/__init__.py]
+parse = __version__ = (?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)
+
```
nitpick: Don't think there should be a blank line here.

AndyMender replied:
Unfortunately, PyCharm (my IDE of choice) maniacally adds newline characters at the end of files. I'll remove this one.

AndyMender replied:
I think I managed to remove this finally.

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+
+if __name__ == "__main__":
+    # bump version globally
+    bump_version()
```
This should probably have some way to specify the part to change via the commandline (e.g. via [argparse](https://docs.python.org/3/library/argparse.html)).

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+    releases = appdata_tree.xpath(version_xpath)[0]
+
+    # attach new release
+    # TODO: use different date string?
```
I guess whatever is in `appdata.xml` currently should be used, not sure if any other formats are supported.

AndyMender replied:
What I meant here is the source of the date string. Should it be the current date or should the string be somehow derived from the git commit?

The-Compiler replied:
Ah! I think the current date (when the release was actually released) is fine.

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,105 @@
+    :return:
+    """
+
+    # create <release> block and populate
```
I don't think those comments add much value - they say the same the code already says. (same for the other comments further below, too).

The-Compiler commented:
I probably also missed some of the stuff that [pylint](https://travis-ci.org/qutebrowser/qutebrowser/jobs/438083039)/[flake8](https://travis-ci.org/qutebrowser/qutebrowser/jobs/438083040) will print - please take a look at those as well. Note that you can also run those locally via `tox -e pylint,flake8`.

AndyMender commented:
I changed the code style where I could and updated qutebrowser.appdata.xml with all past releases since 2014 (extracted from the git tree). I still need some feedback on doc/contributing.asciidoc, since I needed to rebuild the command-line interface in build_release.py to include automatic bumps via update_version.py and the file mentions a download_release.py script, which I could not find in scripts/dev.

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,15 @@ qutebrowser release
-* Commit.
-
-* Create annotated git tag (`git tag -s "v1.$x.$y" -m "Release v1.$x.$y"`).
-* `git push origin`; `git push origin v1.$x.$y`.
```
I don't think `bumpversion` takes care of the pushing?

AndyMender replied:
No, it doesn't, you are correct. It creates the bump commit and the related tag, but doesn't actually push it to remote.

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,15 @@ qutebrowser release
-* Windows: Run `git checkout v1.X.Y; py -3 scripts\dev\build_release.py --asciidoc C:\Python27\python %userprofile%\bin\asciidoc-8.6.10\asciidoc.py --upload v1.X.Y` (replace X/Y by hand).
-* macOS: Run `git checkout v1.X.Y && python3 scripts/dev/build_release.py --upload v1.X.Y` (replace X/Y by hand).
+* Mark the milestone at https://github.com/qutebrowser/qutebrowser/milestones as closed.
+* Linux: run `./.venv/bin/python3 scripts/dev/build_release.py --bump ["major","minor","patch"] --upload`.
```
I don't think it's a good idea to do the bumping as part of `build_release.py` - the GitHub release needs to exist for `build_release.py` to upload the release, and the tag needs to exist to create the release.

Either those two things should be kept separate for now, or you'd need to look into `build_release.py` also pushing the tag and creating the GitHub release via [github3.py](https://github3.readthedocs.io/en/master/).

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,15 @@ qutebrowser release
+* Mark the milestone at https://github.com/qutebrowser/qutebrowser/milestones as closed.
+* Linux: run `./.venv/bin/python3 scripts/dev/build_release.py --bump ["major","minor","patch"] --upload`.
+* Windows: Run `py -3 scripts\dev\build_release.py --asciidoc C:\Python27\python %userprofile%\bin\asciidoc-8.6.10\asciidoc.py -bump ["major","minor","patch"] --upload`.
+* macOS: Run `python3 scripts/dev/build_release.py --bump ["major","minor","patch"] --upload`.
```
Windows/macOS would just pull the commit made on Linux, bumping again there would result in a different commit.

The-Compiler commented on misc/qutebrowser.appdata.xml:
```diff
@@ -41,8 +41,46 @@
 	<url type="donation">https://github.com/qutebrowser/qutebrowser#donating</url>
 	<releases>
-		<release version="1.3.0" date="2018-05-04"/>
+        <release version="1.5.0" date="2018-10-03"/>
```
Looks like there are mixed tabs/spaces here - what does `update_version.py` do?

The-Compiler replied:
note to self: Add 1.5.1 here by hand if it's released by the time I merge this :wink:

The-Compiler commented on qutebrowser/__init__.py:
```diff
@@ -26,8 +26,10 @@
 __description__ = "A keyboard-driven, vim-like browser based on PyQt5."
 
 basedir = os.path.dirname(os.path.realpath(__file__))
+appdata_path = os.path.join(os.path.dirname(basedir), "misc",
```
I'm a bit confused about the `os.path.dirname(basedir)` here - is this just to get to the parent directory? Either way, I think I'd prefer just having this in `update_version.py` as it's not needed in qutebrowser itself.

The-Compiler commented on requirements.txt:
```diff
@@ -4,6 +4,7 @@ attrs==18.2.0
 colorama==0.3.9
 cssutils==1.0.2
 Jinja2==2.10
+lxml==4.2.5
```
Same as with the `bumpversion` requirement also applies here :wink:

The-Compiler commented on scripts/dev/build_release.py:
```diff
@@ -387,10 +400,15 @@ def main():
         input()
-        github_upload(artifacts, args.upload[0])
+
+        # create version tag
```
That comment confused me, as no tag is actually created (as in `git tag`) - I think it'd be best to remove it :wink:

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,102 @@
+        version_leap: define the jump between versions
+        ("major", "minor", "patch")
+    """
+    subprocess.run(['bump2version', version_leap])
```
still needs `check=True` :wink:

The-Compiler commented on scripts/dev/update_version.py:
```diff
@@ -0,0 +1,102 @@
+
+        write_appdata(appdata_tree)
+    else:
+        print("Option 'bump' not specified via command-line."
```
I think it'd be better to require the `bump` argument - after all, there would be no reason to run this script otherwise, and `build_release.py` also doesn't call it in that case.

AndyMender commented:
I feel like there is some "chicken vs egg" confusion in terms of what update_version.py and bump_release.py do currently.

`bumpversion` (or `bump2version`, used at the moment) does the following:
- update the `current_version` param in .bumpversion.cfg
- update `__version__` and indirectly `__version_info__` in `qutebrowser/__init__.py`
- create the related bumping git commit
- create a git tag

`scripts/dev/update_version.py` includes the above procedure, in addition to also adding a new `<release>` block to the `<releases>` parent block in `misc/qutebrowser.appdata.xml`.

I added a wrapper function for `update_version.py` to `build_release.py`, but it is not clear to me when exactly in the pipeline it needs to be executed and how would the new pipeline be explained in `doc/contributing.asciidoc`:
```python
    if args.upload:
        # Fail early when trying to upload without github3 installed
        # or without API token
        import github3  # pylint: disable=unused-variable
        read_github_token()

    if args.bump is not None:
        run_update_version(args)

    if args.no_asciidoc:
        os.makedirs(os.path.join('qutebrowser', 'html', 'doc'), exist_ok=True)
    else:
        run_asciidoc2html(args)

    if os.name == 'nt':
        artifacts = build_windows()
    elif sys.platform == 'darwin':
        artifacts = build_mac()
    else:
        test_makefile()
        artifacts = build_sdist()
        upload_to_pypi = True

    if args.upload:
        from qutebrowser import __version__
        utils.print_title("Press enter to release...")
        input()

        version_tag = "v" + __version__

        github_upload(artifacts, version_tag)
        if upload_to_pypi:
            pypi_upload(artifacts)
    else:
        print()
        utils.print_title("Artifacts")
        for artifact in artifacts:
            print(artifact)
```
The bump itself is also not mandatory in `build_release.py` in case one just wants to build a release without bumping the version.

The-Compiler commented:
The current release workflow roughly works like this:

- Bump version and tag (basically, what `update_version.py` does)
- Push
- Create a release on GitHub (so release artifacts can be uploaded)
- Run `build_release.py` with `--upload` once on Linux/macOS/Windows, so the artifacts are uploaded to GitHub.

To be able to create the GitHub release, the tag needs to be pushed. To be able to run `build_release.py`, the release needs to be created. Also note that `build_release.py` is run three times on three different platforms.

So if `build_release.py` is going to run `update_version.py`, it also needs to create the GitHub release - but it also needs to be able to just build and upload a given version for macOS/Windows.

Alternatively, those two steps should be entirely separate, so I can run `update_version.py` on Linux and then `build_release.py` on Linux/Windows/macOS.

Hope that clears things up?

AndyMender commented:
Yes, everything is much clearer to me now, thank you! I will adjust `contributing.asciidoc` and `build_release.py` accordingly :).

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,18 @@ qutebrowser release
-* macOS: Run `git checkout v1.X.Y && python3 scripts/dev/build_release.py --upload v1.X.Y` (replace X/Y by hand).
+* Mark the milestone at https://github.com/qutebrowser/qutebrowser/milestones as closed.
+* Run `./.venv/bin/python3 scripts/dev/update_version.py major | minor | patch`.
+* Run 'git push origin'.
```
Those should probably be backticks (\`) so those are marked as code.

Also, the tag needs to be pushed individually - only doing `git push origin` won't push the tag.

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,18 @@ qutebrowser release
+* Run `./.venv/bin/python3 scripts/dev/update_version.py major | minor | patch`.
+* Run 'git push origin'.
+* Create new release via GitHub (required to upload release artifacts).
+* Linux: run `./.venv/bin/python3 scripts/dev/build_release.py --upload`.
```
run -> Run

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,18 @@ qutebrowser release
+* Create new release via GitHub (required to upload release artifacts).
+* Linux: run `./.venv/bin/python3 scripts/dev/build_release.py --upload`.
+* Windows: Run `py -3 scripts\dev\build_release.py --asciidoc C:\Python27\python %userprofile%\bin\asciidoc-8.6.10\asciidoc.py --upload`.
+* macOS: Run `python3 scripts/dev/build_release.py --upload`.
```
I think macOS/Windows should still have the `git checkout v1.X.Y` in front - this is to make sure that the commit which was actually tagged also gets released, i.e. `build_release.py` isn't accidentally run on the master branch when e.g. a v1.5.1 patch release is done.

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,24 +696,18 @@ qutebrowser release
 * On server:
   - Run `python3 scripts/dev/download_release.py v1.X.Y` (replace X/Y by hand).
   - Run `git pull github master && sudo python3 scripts/asciidoc2html.py --website /srv/http/qutebrowser`
+* Create release on github.
```
This needs to be done before running `build_release.py` so the release artifacts can be uploaded to GitHub.

AndyMender commented:
I introduced some small tweaks to contributing.asciidoc and update_version.py (now it explicitly returns `__version__` to populate a `VERSION_TAG` variable needed by `git`).

The-Compiler commented on doc/contributing.asciidoc:
```diff
@@ -696,22 +696,15 @@ qutebrowser release
+* Run `git push origin; git push $VERSION_TAG`.
+* Create new release via GitHub (required to upload release artifacts).
+* Linux: Run `git checkout $VERSION_TAG && ./.venv/bin/python3 scripts/dev/build_release.py --upload`.
+* Windows: Run `git checkout $VERSION_TAG; py -3 scripts\dev\build_release.py --asciidoc C:\Python27\python %userprofile%\bin\asciidoc-8.6.10\asciidoc.py --upload`.
```
That's not going to work - `VERSION_TAG` was only set on Linux, so it can't be used directly on Windows/macOS as those are different machines/shells :wink:

What about making `update_version.py` print the appropriate `build_release.py` commands (and removing them from here) instead?

AndyMender replied:
> That's not going to work - VERSION_TAG was only set on Linux, so it can't be used directly on Windows/macOS as those are different machines/shells wink

I knew it looked too simple to be true, eh.

> What about making update_version.py print the appropriate build_release.py commands (and removing them from here) instead?

By that do you mean the entire string of the `build_release.py` command that needs to be run? I think this can be done.

The-Compiler replied:
Yeah, printing the entire commands there instead of having them in `contribution.asciidoc` seems like a nice improvement - then I can copy-paste them from there with the right version numbers.

AndyMender replied:
I moved some of the commands to `update_version.py`, though obviously readability suffered, because they had to be PEP8-compliant. `contributing.asciidoc` was adjusted accordingly.

AndyMender commented:
I introduced changes to `update_version.py` as suggested. Should anything else be improved? :)

The-Compiler commented:
Sorry for the long delay, I didn't really have much time for the v1.5.0 and v1.6.0 releases so I decided to not merge this before those release. I decided to throw myself into the cold water for v1.7.0 now - let's see how that goes!

Thanks for the contribution!

The-Compiler commented:
After pushing some [additional fixes](https://github.com/qutebrowser/qutebrowser/commits/master/scripts/dev/update_version.py) I think this is ready for the v1.7.0 release!

Note that I switched to using bumpversion to update the appdata XML as well - this avoids rewriting the file which loses comments and formatting.
